src/main/scheduler.c

```diff
@@ -0,0 +1,309 @@
+        taskId = currentTaskId;
+
+    if (taskId < TASK_COUNT) {
+        cfTasks[taskId].isEnabled = newEnabledState;
+    }
+}
+
+void scheduler(void)
+{
+    uint8_t taskId;
+    uint8_t selectedTaskId;
+    uint8_t selectedTaskDynPrio;
+
+    static uint32_t timeCpuBusy = 0;
+    static uint32_t timeCpuIdle = 0;
+    static uint32_t previousSchedulerTime = 0;
+    static uint32_t totalTaskExecutionTime = 0;
+
+    /* Cache currentTime */
+    currentTime = micros();
+
+    /* Update task dynamic priorities */
+    for (taskId = 0; taskId < TASK_COUNT; taskId++) {
+        uint16_t newDynamicPriority = 0;
+        uint16_t taskAgeCycles = ((currentTime - cfTasks[taskId].lastExecutedAt) / cfTasks[taskId].desiredPeriod);
+
+        /* Task has checkFunc - event driven */
+        if (cfTasks[taskId].checkFunc != NULL) {
+            /* Increase priority for event driven tasks */
+            if (cfTasks[taskId].checkFunc()) {
+                // FIXME: need cleaner approach here
+                /* If task is already scheduled to execution by event, force time-based scheduling to avoid task starvation */
+                if (cfTasks[taskId].dynamicPriority > 0) {
+                    newDynamicPriority = 1 + cfTasks[taskId].staticPriority * (1 + taskAgeCycles);
+                }
+                else {
+                    newDynamicPriority = 1 + cfTasks[taskId].staticPriority;
+                }
+            }
+        }
+        /* Task is time-driven, dynamicPriority is last execution age measured in desiredPeriods) */
+        else {
+            if (taskAgeCycles) {
+                newDynamicPriority = 1 + cfTasks[taskId].staticPriority * taskAgeCycles;
+            }
+        }
+
+        /* limit new priority to avoid overflow of uint8_t */
+        cfTasks[taskId].dynamicPriority = MIN(newDynamicPriority, 255);
+    }
+
+    /* Find the task with highest dynamicPriority - would be executed next */
+    selectedTaskId = TASK_NONE;
+    selectedTaskDynPrio = 0;
+    for (taskId = 0; taskId < TASK_COUNT; taskId++) {
+        if (cfTasks[taskId].dynamicPriority > selectedTaskDynPrio) {
+            selectedTaskDynPrio = cfTasks[taskId].dynamicPriority;
+            selectedTaskId = taskId;
+        }
+    }
+
+    uint32_t currentTimeAfterTaskCycle;
+
+    /* Found a task that should be run */
+    if (selectedTaskId != TASK_NONE) {
+        uint32_t deltaTime = currentTime - cfTasks[selectedTaskId].lastExecutedAt;
+
+        /* Execute task */
+        currentTaskId = selectedTaskId;
+        if (cfTasks[selectedTaskId].taskFunc != NULL) {
+            cfTasks[selectedTaskId].taskFunc(deltaTime);
+        }
+
+        /* Update task statistics */
+        cfTasks[selectedTaskId].taskLatestDeltaTime = deltaTime;
+        cfTasks[selectedTaskId].taskMaxDeltaTime = MAX(cfTasks[selectedTaskId].taskMaxDeltaTime, deltaTime);
+        cfTasks[selectedTaskId].dynamicPriority = 0;
+        cfTasks[selectedTaskId].lastExecutedAt = currentTime;
```

Comment:
Looks like you take the scheduling time into account here also. Is this only a fraction of the real execution? Shouldn't you measure the execution of taskFunc() only? You could also have dedicated statistics about the scheduler performance.